File: windows/src/FloatingLyric.Core/LyricsProvider.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace FloatingLyric.Core;

public sealed class LyricsProvider(IHttpClient http, ILyricsCache cache, Func<DateTime>? now = null)
{
    public static readonly TimeSpan NegativeTtl = TimeSpan.FromHours(24);

    public const string UserAgent =
        "FloatingLyric/1.0 (Windows lyrics overlay; contact via app repository)";

    /// <summary>How far a candidate's length may sit from the track's before
    /// it stops being the same recording. Widened in stages — see BestMatch.</summary>
    public const double CloseToleranceSeconds = 5.0;
    public const double LooseToleranceSeconds = 15.0;

    private const string Base = "https://lrclib.net";
    private readonly Func<DateTime> _now = now ?? (() => DateTime.UtcNow);

    public async Task<LyricsResult> LyricsAsync(TrackIdentity track)
    {
        if (cache.Read(track.Id, _now()) is { } cached) return cached;
        try
        {
            var result = await FetchAsync(track);
            cache.Write(result, track.Id, _now());
            return result;
        }
        catch
        {
            // Transient failure: report nothing found, but never cache it —
            // otherwise one flaky moment poisons a song for a day.
            return LyricsResult.None;
        }
    }

    private async Task<LyricsResult> FetchAsync(TrackIdentity track) =>
        await DirectLookupAsync(track) ?? await SearchLookupAsync(track);

    private async Task<LyricsResult?> DirectLookupAsync(TrackIdentity track)
    {
        var query = Query(new()
        {
            ["artist_name"] = track.Artist,
            ["track_name"] = track.Title,
            ["album_name"] = track.Album,
            ["duration"] = (track.DurationMs / 1000).ToString(),
        });
        var response = await http.SendAsync(Request($"{Base}/api/get?{query}"));
        if (response.Status is < 200 or >= 300) return null;

        var record = Deserialize<Record>(response.Body);
        return record?.AsResult();
    }

    private async Task<LyricsResult> SearchLookupAsync(TrackIdentity track)
    {
        var query = Query(new()
        {
            ["artist_name"] = track.Artist,
            ["track_name"] = track.Title,
        });
        var response = await http.SendAsync(Request($"{Base}/api/search?{query}"));
        if (response.Status is < 200 or >= 300) return LyricsResult.None;

        var records = Deserialize<List<Record>>(response.Body);
        if (records is null) return LyricsResult.None;

        return BestMatch(track, records)?.AsResult() ?? LyricsResult.None;
    }

    /// <summary>
    /// Picks the best of LRCLIB's search results, widening in stages.
    ///
    /// A remaster, a regional release and a single edit of one song routinely
    /// differ by tens of seconds, and LRCLIB stores several of each. Demanding
    /// a close match throws all of them away and shows the user nothing — when
    /// being seconds out is what the sync offset slider is for.
    /// </summary>
    public static Record? BestMatch(TrackIdentity track, IReadOnlyList<Record> records)
    {
        var wanted = track.DurationMs / 1000.0;
        var usable = records.Where(r => r.AsResult() is not LyricsResult.NotFound).ToList();
        if (usable.Count == 0) return null;

        List<Record> Within(double tolerance) => usable
            .Where(r => r.Duration is { } d && Math.Abs(d - wanted) <= tolerance)
            .ToList();

        // Last resort only: a search for one song can return another, so the
        // widest tier insists the names match. Records with no duration at all
        // reach the user only here.
        var sameSong = usable.Where(r => r.IsSameSong(track)).ToList();

        foreach (var tier in new[] { Within(CloseToleranceSeconds), Within(LooseToleranceSeconds), sameSong })
            if (tier.Count > 0) return Preferred(tier, wanted);

        return null;
    }

    /// <summary>Synced beats plain; then the closest duration wins.</summary>
    private static Record Preferred(List<Record> records, double wanted) => records
        .OrderByDescending(r => r.HasSynced)
        .ThenBy(r => r.DistanceFrom(wanted))
        .First();

    private static HttpRequestMessage Request(string url)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.Add("User-Agent", UserAgent);
        return request;
    }

    private static string Query(Dictionary<string, string> items) => string.Join("&",
        items.Select(kv => $"{Uri.EscapeDataString(kv.Key)}={Uri.EscapeDataString(kv.Value)}"));

    private static T? Deserialize<T>(byte[] body)
    {
        try { return JsonSerializer.Deserialize<T>(body); }
        catch (JsonException) { return default; }
    }

    public sealed record Record
    {
        [JsonPropertyName("duration")] public double? Duration { get; init; }
        [JsonPropertyName("syncedLyrics")] public string? SyncedLyrics { get; init; }
        [JsonPropertyName("plainLyrics")] public string? PlainLyrics { get; init; }
        [JsonPropertyName("artistName")] public string? ArtistName { get; init; }
        [JsonPropertyName("trackName")] public string? TrackName { get; init; }

        public LyricsResult AsResult()
        {
            if (!string.IsNullOrEmpty(SyncedLyrics))
            {
                var lines = LrcParser.Parse(SyncedLyrics);
                if (lines.Count > 0) return new LyricsResult.Synced(lines);
            }
            if (!string.IsNullOrEmpty(PlainLyrics)) return new LyricsResult.Plain(PlainLyrics);
            return LyricsResult.None;
        }

        public bool HasSynced => AsResult() is LyricsResult.Synced;

        /// <summary>No duration is infinitely far away, so it sorts last
        /// wherever anything better exists.</summary>
        public double DistanceFrom(double seconds) =>
            Duration is { } d ? Math.Abs(d - seconds) : double.MaxValue;

        public bool IsSameSong(TrackIdentity track) =>
            Normalized(ArtistName) == Normalized(track.Artist) &&
            Normalized(TrackName) == Normalized(track.Title);

        private static string Normalized(string? text) => (text ?? "")
            .Trim()
            .Normalize(System.Text.NormalizationForm.FormD)
            .Where(c => System.Globalization.CharUnicodeInfo.GetUnicodeCategory(c)
                        != System.Globalization.UnicodeCategory.NonSpacingMark)
            .Aggregate(new System.Text.StringBuilder(), (sb, c) => sb.Append(c))
            .ToString()
            .ToLowerInvariant();
    }
}

File: macos/Sources/FloatingLyricCore/Lyrics/LyricsProvider.swift
import Foundation

public final class LyricsProvider: @unchecked Sendable {
    public static let negativeTTL: TimeInterval = 86_400
    public static let userAgent =
        "FloatingLyric/1.0 (macOS lyrics overlay; contact via app repository)"
    /// How far a candidate's duration may sit from the track's before it stops
    /// being the same recording. Widened in stages rather than all at once —
    /// see `bestMatch(for:in:)`.
    static let closeToleranceSeconds = 5.0
    static let looseToleranceSeconds = 15.0
    private static let base = "https://lrclib.net"

    private let http: HTTPClient
    private let cache: LyricsCaching
    private let now: () -> Date

    public init(http: HTTPClient, cache: LyricsCaching, now: @escaping () -> Date = Date.init) {
        self.http = http
        self.cache = cache
        self.now = now
    }

    public func lyrics(for track: TrackIdentity) async -> LyricsResult {
        if let cached = cache.read(trackID: track.id, now: now()) { return cached }
        do {
            let result = try await fetch(track)
            cache.write(result, trackID: track.id, now: now())
            return result
        } catch {
            // Transient failure: report nothing found, but never cache it.
            return .notFound
        }
    }

    private func fetch(_ track: TrackIdentity) async throws -> LyricsResult {
        if let direct = try await directLookup(track) { return direct }
        return try await searchLookup(track)
    }

    private func directLookup(_ track: TrackIdentity) async throws -> LyricsResult? {
        var components = URLComponents(string: Self.base + "/api/get")!
        components.queryItems = [
            .init(name: "artist_name", value: track.artist),
            .init(name: "track_name", value: track.title),
            .init(name: "album_name", value: track.album),
            .init(name: "duration", value: String(track.durationMs / 1000)),
        ]
        let response = try await http.send(request(components.url!))
        guard (200..<300).contains(response.status) else { return nil }
        guard let record = try? JSONDecoder().decode(Record.self, from: response.body)
        else { return nil }
        return record.asResult
    }

    private func searchLookup(_ track: TrackIdentity) async throws -> LyricsResult {
        var components = URLComponents(string: Self.base + "/api/search")!
        components.queryItems = [
            .init(name: "artist_name", value: track.artist),
            .init(name: "track_name", value: track.title),
        ]
        let response = try await http.send(request(components.url!))
        guard (200..<300).contains(response.status),
              let records = try? JSONDecoder().decode([Record].self, from: response.body)
        else { return .notFound }

        return Self.bestMatch(for: track, in: records)?.asResult ?? .notFound
    }

    /// Picks the best of LRCLIB's search results, widening in stages.
    ///
    /// A remaster, a regional release and a single edit of the same song
    /// routinely differ by more than a few seconds, and LRCLIB stores several
    /// of each. Demanding a close duration match throws all of them away and
    /// leaves the user with nothing — where being seconds out is something the
    /// sync offset slider already fixes. So: an exact-ish duration if one
    /// exists, then a loose one, and only then any entry that is unambiguously
    /// the same song by name.
    static func bestMatch(for track: TrackIdentity, in records: [Record]) -> Record? {
        let wanted = Double(track.durationMs) / 1000
        let usable = records.filter { $0.asResult != .notFound }
        guard !usable.isEmpty else { return nil }

        func within(_ tolerance: Double) -> [Record] {
            usable.filter { record in
                guard let duration = record.duration else { return false }
                return abs(duration - wanted) <= tolerance
            }
        }

        // Last resort only: a search for one song can return another, so the
        // widest tier insists the names actually match. Records with no
        // duration at all reach the user only here.
        let sameSong = usable.filter { $0.isSameSong(as: track) }

        for tier in [within(Self.closeToleranceSeconds),
                     within(Self.looseToleranceSeconds),
                     sameSong] where !tier.isEmpty {
            return preferred(in: tier, wanted: wanted)
        }
        return nil
    }

    /// Synced lyrics beat plain ones; after that, the closest duration wins.
    private static func preferred(in records: [Record], wanted: Double) -> Record? {
        records.min { a, b in
            if a.hasSynced != b.hasSynced { return a.hasSynced }
            return a.distance(from: wanted) < b.distance(from: wanted)
        }
    }

    private func request(_ url: URL) -> URLRequest {
        var request = URLRequest(url: url)
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        return request
    }

    struct Record: Decodable {
        let duration: Double?
        let syncedLyrics: String?
        let plainLyrics: String?
        let artistName: String?
        let trackName: String?

        var asResult: LyricsResult {
            if let synced = syncedLyrics, !synced.isEmpty {
                let lines = LRCParser.parse(synced)
                if !lines.isEmpty { return .synced(lines) }
            }
            if let plain = plainLyrics, !plain.isEmpty { return .plain(plain) }
            return .notFound
        }

        var hasSynced: Bool {
            if case .synced = asResult { return true }
            return false
        }

        /// A record with no duration is infinitely far away, so it sorts last
        /// wherever something better exists.
        func distance(from seconds: Double) -> Double {
            guard let duration else { return .greatestFiniteMagnitude }
            return abs(duration - seconds)
        }

        func isSameSong(as track: TrackIdentity) -> Bool {
            Record.normalized(artistName) == Record.normalized(track.artist)
                && Record.normalized(trackName) == Record.normalized(track.title)
        }

        private static func normalized(_ text: String?) -> String {
            (text ?? "").folding(options: [.caseInsensitive, .diacriticInsensitive],
                                 locale: nil)
                .trimmingCharacters(in: .whitespacesAndNewlines)
        }
    }
}
